### ISSUE TITLE
New rule: disallow nested es6 destructuring

### ISSUE THREAD

**Issue by mik01aj** [eslint/eslint#6722]
The rule should warn in these scenarios:

```
let {a: {b}, c} = obj;    // 1
let {a, a: {b}, c} = obj; // 2
let [a, {b}] = arr;       // 3
let [[a]] = arr;          // 4
```

The rule should NOT warn in these scenarios:

```
let {c} = obj; let {b} = obj.a; // refactored 1
let {a, c} = obj; let {b} = a;  // refactored 2
let {a, b, c} = obj;
let [a, b, c] = arr;
```

This rule is stylistic. Destructuring of nested objects can be confusing and is usually better written in several smaller steps.

I think it's generic enough for core.

I'm thinking also about configuration options to ignore array destructuring (then examples 3 and 4 would be valid).

I don't think I can implement it anytime near, but if I need it urgently somewhere, I will do :)

**Comment by platinumazure** [comment]
Proposal looks complete except for the configuration options to ignore array destructuring.

I do have one question about ignoring array destructure:

```js
let [a, {b: {c}}] = arr;       // nested destructure in array (modified from example 3)
```

Should that be flagged? Can you add a few more examples along those lines to show what would be valid or invalid with the configuration option?

**Comment by mik01aj** [comment]
with `{allowArrays: true}` it should still warn in these scenarios:

```
let [a, {b: {c}}] = arr;
let {a: [{b}]} = obj;
let {a: [{b: c}]} = obj;
```

but it should NOT warn in these ones (these would be invalid without the option):

```
let [[[[a]]]] = arr;
let [a, [b]] = arr;
let [a, {b}] = arr;
let [...a, {b}] = arr;
let {a: [b]} = obj;
```

**Comment by nzakas** [comment]
@mik01aj thanks for the submission. Please take a moment to read over our guide on proposing new rules: http://eslint.org/docs/developer-guide/contributing/new-rules

Specifically, we do require proposers to implement their suggestions if accepted (we just can't be a rule factory for everyone's ideas), so your willingness to implement it is a key decision point.

My own evaluation: I don't see this as flagging an error or implementing part of a popular style guide (as mentioned in the guide), so I don't think this is a good core rule.

**Comment by kaicataldo** [comment]
I'm not entirely sure this should be in core either, but just a thought: what if the rule allowed for more fine-grained control over how many levels deep you can destructure?

**Comment by IanVS** [comment]
:+1: Personally I've spent way too much time trying to decipher nested destructurings, and would love to see a rule like this.  Whether it's in core, or a plugin is up for debate, but I think it would be a great rule regardless.

**Comment by nzakas** [comment]
@ianvs the issue here is whether or not it belongs in core, not necessarily if the rule is useful. Remember, we purposely raised the bar for the core because we already have over 200 rules.

And a reminder for everyone: our job is not to determine if a rule is useful, it's to determine if it's so important that ESLint would seem incomplete without it. Our guidelines: http://eslint.org/docs/developer-guide/contributing/new-rules

**Comment by alberto** [comment]
There was also discussion about nested destructuring at: https://github.com/eslint/eslint/issues/6053

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
